File: forms/medical-language-speaking-assessment-for-english/back-end-with-loco/tests/engine/oet_grader_test.rs
use medical_language_speaking_assessment_for_english_loco_crate::engine::oet_grader::grade;
use medical_language_speaking_assessment_for_english_loco_crate::engine::rules::criterion_registry;
use medical_language_speaking_assessment_for_english_loco_crate::engine::types::*;

/// Build a fully-rated assessment with perfect (max-band) scores.
fn perfect_case() -> AssessmentData {
    AssessmentData {
        candidate: CandidateDetails {
            candidate_id: "C-001".to_string(),
            candidate_name: "Anika Sharma".to_string(),
            examiner_name: "Dr Examiner".to_string(),
            test_centre: "London".to_string(),
            test_date: "2026-06-01".to_string(),
            profession: "medicine".to_string(),
            first_language: "Hindi".to_string(),
            country_of_training: "India".to_string(),
            years_of_experience: "5".to_string(),
        },
        role_play1: RolePlayContext {
            scenario_title: "Headache follow-up".to_string(),
            scenario_summary: "Patient interview about persistent headaches.".to_string(),
            patient_role: "Adult patient".to_string(),
            setting: "GP clinic".to_string(),
            safety_criticality: "standard".to_string(),
            examiner_notes: String::new(),
        },
        role_play2: RolePlayContext {
            scenario_title: "Discharge plan".to_string(),
            scenario_summary: "Explain medication and follow-up.".to_string(),
            patient_role: "Adult patient".to_string(),
            setting: "Outpatient".to_string(),
            safety_criticality: "standard".to_string(),
            examiner_notes: String::new(),
        },
        linguistic_role_play1: LinguisticRating {
            intelligibility: Some(6.0),
            fluency: Some(6.0),
            appropriateness_of_language: Some(6.0),
            resources_of_grammar_and_expression: Some(6.0),
        },
        linguistic_role_play2: LinguisticRating {
            intelligibility: Some(6.0),
            fluency: Some(6.0),
            appropriateness_of_language: Some(6.0),
            resources_of_grammar_and_expression: Some(6.0),
        },
        clinical_indicators: ClinicalIndicators {
            relationship_building: Some(3.0),
            understanding_patient_perspective: Some(3.0),
            providing_structure: Some(3.0),
            information_gathering: Some(3.0),
            information_giving: Some(3.0),
            examiner_notes: String::new(),
        },
    }
}

#[test]
fn empty_assessment_grades_as_e() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.grade, "E");
    assert_eq!(result.scaled_score, 0);
    assert_eq!(result.linguistic_total, 0.0);
    assert_eq!(result.clinical_total, 0.0);
    assert_eq!(result.raw_total, 0.0);
}

#[test]
fn perfect_assessment_grades_as_a() {
    let data = perfect_case();
    let result = grade(&data);
    assert_eq!(result.grade, "A");
    assert_eq!(result.scaled_score, 500);
    assert_eq!(result.linguistic_total, 24.0);
    assert_eq!(result.clinical_total, 15.0);
    assert_eq!(result.raw_total, 39.0);
}

#[test]
fn criterion_ids_are_unique_and_complete() {
    let registry = criterion_registry();
    assert_eq!(registry.len(), 9, "9 criteria: 4 linguistic + 5 clinical");
    let mut ids: Vec<&str> = registry.iter().map(|c| c.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "criterion IDs must be unique");
}

#[test]
fn linguistic_mean_or_single_uses_single_value() {
    // Only role-play 1 rated; the engine should treat it as the mean.
    let mut data = perfect_case();
    data.linguistic_role_play2 = LinguisticRating::default();
    data.clinical_indicators = ClinicalIndicators::default();
    let result = grade(&data);
    assert_eq!(result.linguistic_total, 24.0);
    // Without clinical ratings the clinical total stays at 0.
    assert_eq!(result.clinical_total, 0.0);
}

#[test]
fn per_criterion_table_has_nine_entries() {
    let data = perfect_case();
    let result = grade(&data);
    assert_eq!(result.per_criterion_scores.len(), 9);
}

#[test]
fn scaled_score_band_boundaries() {
    // Half of perfect (linguistic only) ≈ 308 -> grade C+
    let mut data = perfect_case();
    data.clinical_indicators = ClinicalIndicators::default();
    let result = grade(&data);
    // 24/39 * 500 = 307.69 -> 308
    assert_eq!(result.scaled_score, 308);
    assert_eq!(result.grade, "C+");
}

#[test]
fn fired_rules_only_for_rated_criteria() {
    let mut data = perfect_case();
    // Drop one criterion across both role-plays so it is unrated.
    data.linguistic_role_play1.fluency = None;
    data.linguistic_role_play2.fluency = None;
    let result = grade(&data);
    assert!(result.fired_rules.iter().any(|r| r.id == "LING-INT"));
    assert!(!result.fired_rules.iter().any(|r| r.id == "LING-FLU"));
}

File: forms/medical-language-speaking-assessment-for-cymraeg/front-end-with-svelte/src/lib/engine/oet-grader.test.ts
import { describe, it, expect } from 'vitest';
import { gradeOET, gradeAssessment, classifyScaledScore } from './oet-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { criterionRegistry } from './rules';
import type { AssessmentData } from './types';

/** A blank assessment (mirrors the store's createDefaultAssessment, without the `$app` import). */
function createDefaultAssessment(): AssessmentData {
	return {
		candidate: {
			candidateId: '',
			candidateName: '',
			examinerName: '',
			testCentre: '',
			testDate: '',
			profession: 'medicine',
			firstLanguage: '',
			countryOfTraining: '',
			yearsOfExperience: ''
		},
		rolePlay1: {
			scenarioTitle: '',
			scenarioSummary: '',
			patientRole: '',
			setting: '',
			safetyCriticality: '',
			examinerNotes: ''
		},
		rolePlay2: {
			scenarioTitle: '',
			scenarioSummary: '',
			patientRole: '',
			setting: '',
			safetyCriticality: '',
			examinerNotes: ''
		},
		linguisticRolePlay1: { fluency: null, grammar: null, pronunciation: null, clinicalAppropriateness: null },
		linguisticRolePlay2: { fluency: null, grammar: null, pronunciation: null, clinicalAppropriateness: null },
		clinicalIndicators: {
			relationshipBuilding: null,
			understandingPatientPerspective: null,
			providingStructure: null,
			informationGathering: null,
			informationGiving: null,
			examinerNotes: ''
		}
	};
}

/** A top-band candidate: all linguistic 6, all clinical 3. */
function expertCandidate(): AssessmentData {
	const d = createDefaultAssessment();
	d.candidate.candidateName = 'Eleri Top';
	d.linguisticRolePlay1 = { fluency: 6, grammar: 6, pronunciation: 6, clinicalAppropriateness: 6 };
	d.linguisticRolePlay2 = { fluency: 6, grammar: 6, pronunciation: 6, clinicalAppropriateness: 6 };
	d.clinicalIndicators = {
		...d.clinicalIndicators,
		relationshipBuilding: 3,
		understandingPatientPerspective: 3,
		providingStructure: 3,
		informationGathering: 3,
		informationGiving: 3
	};
	return d;
}

/** A bottom-band candidate: linguistic 2, clinical 1. */
function limitedCandidate(): AssessmentData {
	const d = createDefaultAssessment();
	d.linguisticRolePlay1 = { fluency: 2, grammar: 2, pronunciation: 2, clinicalAppropriateness: 2 };
	d.linguisticRolePlay2 = { fluency: 2, grammar: 2, pronunciation: 2, clinicalAppropriateness: 2 };
	d.clinicalIndicators = {
		...d.clinicalIndicators,
		relationshipBuilding: 1,
		understandingPatientPerspective: 1,
		providingStructure: 1,
		informationGathering: 1,
		informationGiving: 1
	};
	return d;
}

describe('classifyScaledScore', () => {
	it('maps cut-points to grades', () => {
		expect(classifyScaledScore(500)).toBe('A');
		expect(classifyScaledScore(450)).toBe('A');
		expect(classifyScaledScore(440)).toBe('B');
		expect(classifyScaledScore(350)).toBe('B');
		expect(classifyScaledScore(340)).toBe('C+');
		expect(classifyScaledScore(300)).toBe('C+');
		expect(classifyScaledScore(290)).toBe('C');
		expect(classifyScaledScore(200)).toBe('C');
		expect(classifyScaledScore(190)).toBe('D');
		expect(classifyScaledScore(100)).toBe('D');
		expect(classifyScaledScore(90)).toBe('E');
		expect(classifyScaledScore(0)).toBe('E');
	});
});

describe('gradeOET', () => {
	it('grades an expert candidate as A with a perfect score', () => {
		const r = gradeOET(expertCandidate());
		expect(r.linguisticTotal).toBe(24);
		expect(r.clinicalTotal).toBe(15);
		expect(r.rawTotal).toBe(39);
		expect(r.scaledScore).toBe(500);
		expect(r.grade).toBe('A');
	});

	it('grades a limited candidate in the lower bands', () => {
		const r = gradeOET(limitedCandidate());
		expect(r.linguisticTotal).toBe(8);
		expect(r.clinicalTotal).toBe(5);
		expect(r.rawTotal).toBe(13);
		expect(r.scaledScore).toBe(167);
		expect(r.grade).toBe('D');
	});

	it('grades a fully blank assessment as E', () => {
		const r = gradeOET(createDefaultAssessment());
		expect(r.rawTotal).toBe(0);
		expect(r.scaledScore).toBe(0);
		expect(r.grade).toBe('E');
	});

	it('averages the two role-plays for linguistic criteria', () => {
		const d = createDefaultAssessment();
		d.linguisticRolePlay1 = { fluency: 6, grammar: 6, pronunciation: 6, clinicalAppropriateness: 6 };
		d.linguisticRolePlay2 = { fluency: 4, grammar: 4, pronunciation: 4, clinicalAppropriateness: 4 };
		const r = gradeOET(d);
		const flu = r.perCriterionScores.find((s) => s.id === 'LING-FLU');
		expect(flu?.meanScore).toBe(5);
		expect(r.linguisticTotal).toBe(20);
	});

	it('has a unique id per criterion', () => {
		const ids = criterionRegistry.map((c) => c.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('detectAdditionalFlags', () => {
	it('flags a high grade-band concern for limited candidates', () => {
		const data = limitedCandidate();
		const core = gradeOET(data);
		const flags = detectAdditionalFlags(data, core);
		expect(flags.some((f) => f.id === 'FLAG-GRADE-001')).toBe(true);
		expect(flags.some((f) => f.id === 'FLAG-LING-PRO-001')).toBe(true);
	});

	it('raises no grade-band flag for an expert candidate', () => {
		const data = expertCandidate();
		const core = gradeOET(data);
		const flags = detectAdditionalFlags(data, core);
		expect(flags.some((f) => f.category === 'Overall Grade')).toBe(false);
	});

	it('escalates a clinical failure to high in a safety-critical scenario', () => {
		const data = expertCandidate();
		data.rolePlay1.safetyCriticality = 'high';
		data.clinicalIndicators.informationGiving = 0;
		const core = gradeOET(data);
		const flags = detectAdditionalFlags(data, core);
		const f = flags.find((x) => x.id === 'FLAG-CLIN-informationGiving-0');
		expect(f?.priority).toBe('high');
	});

	it('sorts flags by priority (high first)', () => {
		const data = limitedCandidate();
		data.rolePlay1.examinerNotes = 'Some notes';
		const core = gradeOET(data);
		const flags = detectAdditionalFlags(data, core);
		const order = { high: 0, medium: 1, low: 2 } as const;
		const priorities = flags.map((f) => order[f.priority]);
		expect(priorities).toEqual([...priorities].sort((a, b) => a - b));
	});
});

describe('gradeAssessment', () => {
	it('attaches flags and a timestamp', () => {
		const r = gradeAssessment(limitedCandidate());
		expect(Array.isArray(r.additionalFlags)).toBe(true);
		expect(typeof r.timestamp).toBe('string');
		expect(r.grade).toBe('D');
	});
});
